lib/fluent/plugin/filter_grep.rb

```diff
@@ -54,65 +59,139 @@ def initialize
+        raise Fluent::ConfigError, "exclude#{i} contains a duplicated key, #{key}" if @_exclude_or_conditions[key]
+        @_exclude_or_conditions[key] = Expression.new(record_accessor_create(key), Regexp.compile(exclude))
       end
 
       @regexps.each do |e|
-        raise Fluent::ConfigError, "Duplicate key: #{e.key}" if rs.key?(e.key)
-        rs[e.key] = e.pattern
+        raise Fluent::ConfigError, "Duplicate key: #{e.key}" if @_regexp_and_conditions.key?(e.key)
+        @_regexp_and_conditions[e.key] = Expression.new(record_accessor_create(e.key), e.pattern)
       end
       @excludes.each do |e|
-        raise Fluent::ConfigError, "Duplicate key: #{e.key}" if es.key?(e.key)
-        es[e.key] = e.pattern
+        raise Fluent::ConfigError, "Duplicate key: #{e.key}" if @_exclude_or_conditions.key?(e.key)
+        @_exclude_or_conditions[e.key] = Expression.new(record_accessor_create(e.key), e.pattern)
       end
-
-      rs.each_pair do |k, v|
-        @_regexps[record_accessor_create(k)] = v
+      @and_conditions.each do |and_condition|
+        if !and_condition.regexps.empty? && !and_condition.excludes.empty?
+          raise Fluent::ConfigError, "Do not specify both <regexp> and <exclude> in <and>"
+        end
+        and_condition.regexps.each do |e|
+          raise Fluent::ConfigError, "Duplicate key in <and>: #{e.key}" if @_regexp_and_conditions.key?(e.key)
+          @_regexp_and_conditions[e.key] = Expression.new(record_accessor_create(e.key), e.pattern)
+        end
+        and_condition.excludes.each do |e|
+          raise Fluent::ConfigError, "Duplicate key in <and>: #{e.key}" if @_exclude_and_conditions.key?(e.key)
+          @_exclude_and_conditions[e.key] = Expression.new(record_accessor_create(e.key), e.pattern)
+        end
       end
-      es.each_pair do |k, v|
-        @_excludes[record_accessor_create(k)] = v
+
+      @or_conditions.each do |or_condition|
+        if !or_condition.regexps.empty? && !or_condition.excludes.empty?
+          raise Fluent::ConfigError, "Do not specify both <regexp> and <exclude> in <or>"
+        end
+        or_condition.regexps.each do |e|
+          raise Fluent::ConfigError, "Duplicate key in <or>: #{e.key}" if @_regexp_or_conditions.key?(e.key)
+          @_regexp_or_conditions[e.key] = Expression.new(record_accessor_create(e.key), e.pattern)
+        end
+        or_condition.excludes.each do |e|
+          raise Fluent::ConfigError, "Duplicate key in <or>: #{e.key}" if @_exclude_or_conditions.key?(e.key)
+          @_exclude_or_conditions[e.key] = Expression.new(record_accessor_create(e.key), e.pattern)
+        end
       end
     end
 
     def filter(tag, time, record)
       result = nil
       begin
         catch(:break_loop) do
-          @_regexps.each do |key, regexp|
-            throw :break_loop unless ::Fluent::StringUtil.match_regexp(regexp, key.call(record).to_s)
+          @_regexp_and_conditions.each do |key, expression|
```

Comment:
This is very small optimization but how about using array for conditions?
In `filter`, we don't use `key` and hash related operations so array is good structure.